src/app.ts

```diff
@@ -24,12 +24,72 @@ import ServicesController from "./controllers/servicesController";
 
 import { Express } from "express";
 import expressEnforcesSsl = require("express-enforces-ssl");
+import { Either } from "fp-ts/lib/Either";
 import {
   EnvironmentNodeEnv,
   EnvironmentNodeEnvEnum
 } from "./types/environment";
 import { User } from "./types/user";
 
+/**
+ * Return a response to the client.
+ */
+function respond(
+  response: Either<Error, IResponse>,
+  res: express.Response
+): void {
+  response.fold(
+    err => {
+      res.status(500).json({ message: err.message });
+    },
+    data => {
+      if (data.status === 301) {
+        res.redirect(data.body);
+      } else {
+        res.json(data.body);
+      }
+    }
+  );
+}
+
+/**
+ * Catch SPID authentication errors and redirect the client to
+ * clientErrorRedirectionUrl.
+ */
+const withSpidAuth = (
+  controller: AuthenticationController,
+  clientErrorRedirectionUrl: string,
+  clientLoginRedirectionUrl: string
+): ((
+  req: express.Request,
+  res: express.Response,
+  next: express.NextFunction
+) => void) => {
+  return (
+    req: express.Request,
+    res: express.Response,
+    next: express.NextFunction
+  ) => {
+    passport.authenticate("spid", async (err, user: User) => {
+      if (err) {
+        res.redirect(clientErrorRedirectionUrl);
+        return;
+      }
+      if (!user) {
+        return res.redirect(clientLoginRedirectionUrl);
+      }
+      const maybeResponse = await controller.acs(user);
+      respond(maybeResponse, res);
+    })(req, res, next);
+  };
+};
+
+export interface IResponse {
```

Comment:
`any` should really be the last resort - I suggest to improve `IResponse` by making it parametrized (making it possible to express what gets returned in the `body`) - it's similar to what we're doing in the APIs:

```ts
export interface IResponse<T> {
  readonly body: T;
  readonly status: number;
}
```